### PR TITLE
Don't allow unsafe script mixing in identifiers (UTS 39, C3)

### PR DESCRIPTION
**Pull request by mrluc**
Prevents unsafe script mixing in identifiers. This catches issues in code that look innocent but could be confusing:

```
iex> if аdmin, do: :ok, else: :err
** (SyntaxError) iex:1:4: Unsafe mixed-script identifier 'аdmin' using {Cyrillic, Latin} does not 
meet the requirements of Unicode TS39 Section 5.2, 'Highly Restrictive'.

  \u0430 'а' {Cyrillic}
  \u0064 'd' {Latin}
  \u006D 'm' {Latin}
  \u0069 'i' {Latin}
  \u006E 'n' {Latin}

аdmin
    |
  1 | if аdmin, do: :ok, else: :err
    |    ^
```

We implement the definition of 'Highly Restrictive' for identifiers from UTS 39 Section 5.2. This requires implementation of 5.1 (to detect if scripts resolve, before then checking if they are covered by the specified exceptions), and implies conformance with the Clause 3 capability, 'Mixed Script Detection'.

Preventing script mixing in a restrictive way also has the benefit of preventing mixed-script confusable characters/variations, which is a step towards the recommendation from [trojan source](https://trojansource.codes):

> Compilers, interpreters, and build pipelines supporting Unicode should throw errors or warnings for unterminated bidirectional control characters in comments or string literals, and for **identifiers with mixed-script confusable characters**.

In case you're familiar with Rust's use of the C3 capabilities to protect against unsafe or suspicious uses of mixed scripts in a 'character-oriented' way:

* this PR uses the same core UTS 39 'Mixed-Script Detection' capabilities that the unicode-rs and unicode-security crates provide, but in a different way than the non_ascii_idents lint in Rust uses them; a breakdown of the tradeoffs and ways that either of the two approaches can each be 'stronger' is [here](https://github.com/elixir-lang/elixir/pull/11582#issuecomment-1027452714).
* A character-oriented check could still be added to Elixir in the future; the original [reference implementation](https://github.com/elixir-lang/elixir/pull/11569) includes an example
* But there's also the possibility that other changes could work best, like [letting users define acceptable scripts in their mix.exs](https://github.com/elixir-lang/elixir/pull/11569#issuecomment-1033864019) which would remove the need to worry that only confusable characters from a script have been introduced; rather, no additional script could ever be added 'by surprise' to a project).

This is a Draft because of ongoing discussion -- and, for instance, I went with throwing an error instead of warning because it was simplest, and whether or not that's acceptable may still be an open question!

### CODE REVIEW

**Comment by mrluc** [comment]
@josevalim the CI fails on something that's right at the intersection of these questions, heh -- we like the microsecond character in Elixir, but this one is Greek! (Which is, of course, a scriptset that has many characters confusable with English).

There's a non-Greek, mathematical/technical character for that `μ` symbol though, which would be more correct, and which you can type with Opt + m on OSX. Currently, however, that character is not Allowed by the General Security Profile; that's where having a whitelist of categories or individual characters in the calculation of the 'not-allowed' character list would come in.

```
mix format failed for file: lib/elixir/test/elixir/config/provider_test.exs
** (SyntaxError) lib/elixir/test/elixir/config/provider_test.exs:85:43: Unsafe mixed-script 
identifier 'time_μs' using {Greek, Latin} does not meet the requirements of Unicode 
TS39 Section 5.2, 'Highly Restrictive'.

  \u0074 't' {Latin}
  \u0069 'i' {Latin}
  \u006D 'm' {Latin}
  \u0065 'e' {Latin}
  \u005F '_' {Common}
  \u03BC 'μ' {Greek}
  \u0073 's' {Latin}

time_μs
    |
 85 |     @tag sys_config: [my_app: [encoding: {:time_μs, :"£", "£", '£'}]]
    |                                           ^
```

File: lib/elixir/unicode/security.ex [hunk 1]
```diff
@@ -108,4 +124,252 @@ defmodule String.Tokenizer.Security do
...
+  defp continue([c | chars], sets), do: continue(chars, add(sets, scripts(c)))
+  defp continue([], sets), do: sets
+
+  # - 'ignore chars w/Extensions {Common}, {Inherited} when testing for diffs in script'
+  # - 'don't resolve {ø} for all-common identifiers, like "_123"' (Table 1a)
+  defp add({resolved, seen} = sets, ss) do
+    cond do
+      @empty == seen -> {ss, ss}
+      all?(ss) -> sets
+      not all?(ss) and (all?(seen) or seen == @empty) -> {ss, ss}
```
Comment (josevalim) [inline lib/elixir/unicode/security.ex:292]: We should be able to remove the and:

```suggestion
      all?(seen) or seen == @empty -> {ss, ss}
```

and then I think we should be able to remove the or:

```suggestion
      all?(seen) -> {ss, ss}
```

My concern however is that there is too much logic in the module body that we cannot test directly. My suggestion is for us to expose the `resolve` function so we can unit test. I will have more thoughts shortly. :)

**Comment by josevalim** [comment]
Hi @mrluc!

This has been extremely helpful as always! :heart: 

I have started integrating the code to the tokenizer. The main reason is because we only tokenize a subset of the character set so restricting the resolved set computation to said character set reduces the bytecode considerably. You can see the WIP in this branch: https://github.com/elixir-lang/elixir/tree/uts39-c3

It only adds 2kb to the tokenizer which is very helpful. The impact on tokenization is also very low (one `band` operation per character). I also don't count/keep `all` characters at all (they are given a "top" scriptset which is the union of all). Once that is ready for review, I will submit it here too on top of your PR!

**Comment by josevalim** [comment]
Closing this in favor of #11621.

**Comment by mrluc** [comment]
> I have started integrating the code to the tokenizer.

SO COOL! 🥳 That does seem like the long-term best way. Just took a scan over it, and I literally laughed out loud a couple of times, **tons** of improvements, will put more comments there.

> The main reason is because we only tokenize a subset of the character set so restricting the resolved set computation to said character set reduces the bytecode considerably. You can see the WIP in this branch: https://github.com/elixir-lang/elixir/tree/uts39-c3

Great -- I did that worse in ref impl and then deleted it, but never fixed back up. (Originally it called back into Tokenizer at compile-time to see if codepoints would tokenize -- which had arrows going the wrong way, and thus became impossible w the bootstrapped load order once security.ex was broken out. It didn't occur to me that that all script/mixed-script stuff just 'belongs' in the tokenizer though).

> I also don't count/keep all characters at all (they are given a "top" scriptset which is the union of all)

Oh, nice! Scriptset will never be empty that way. I'm having fun following along w/that other branch.

**Pull request closed without merging**